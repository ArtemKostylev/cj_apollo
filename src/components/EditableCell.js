import React from "react";
import { AVAILABLE_MARKS } from "../scripts/constants";
import { useOnClickOutside } from "../scripts/utils";
import "../styles/Journal.css";

export const EditableCell = ({
  value: initialValue,
  row: index,
  column: id,
  weekend,
  updateMyData,
  group = -1,
}) => {
  const [value, setValue] = React.useState(initialValue);
  const [dropdownVisible, setDropdownVisibility] = React.useState(false);
  const [dropdownWidth, setDropdownWidth] = React.useState(0);

  const Dropdown = (visible) => {
    const onClick = (e) => {
      if (AVAILABLE_MARKS.includes(e.target.innerHTML)) {
        let mark = "";
        if (e.target.innerHTML !== "Пусто") mark = e.target.innerHTML;
        setValue(mark);
        setDropdownVisibility(false);
        if (!updateMyData(index, id, mark, group)) setValue("");
      }
    };

    return (
      <div
        className={`dropdown ${visible ? "visible" : ""}`}
        style={{ width: `${dropdownWidth}px` }}
      >
        <ul>
          <li onClick={onClick}>Пусто</li>
          <li onClick={onClick}>.</li>
          <li onClick={onClick}>2</li>
          <li onClick={onClick}>3</li>
          <li onClick={onClick}>4</li>
          <li onClick={onClick}>5</li>
          <li onClick={onClick}>Б</li>
        </ul>
      </div>
    );
  };

  const ref = React.useRef();

  useOnClickOutside(ref, () => setDropdownVisibility(false));

  const onClick = () => {
    if (!dropdownVisible) setDropdownVisibility(true);
  };

  React.useEffect(() => {
    setValue(initialValue);
  }, [initialValue]);

  React.useEffect(() => {
    setDropdownWidth(ref.current.clientWidth);
  }, [setDropdownWidth]);

  return (
    <td onClick={onClick} ref={ref} className={`${weekend}`}>
      <p className="input">{value}</p>
      {Dropdown(dropdownVisible)}
    </td>
  );
};
