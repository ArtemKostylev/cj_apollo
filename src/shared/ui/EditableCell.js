import React, { useState, useEffect, useRef } from 'react';
import styled from 'styled-components';
import { useOnClickOutside } from '../../utils/utils';
import { Dropdown } from './Dropdown';

const CellText = styled.p`
  border: none;
  outline: none;
  width: 100%;
  height: 100%;
  margin: 0px;
`;

const EditableCellLayout = styled.td`
  border-top: none;
  border: 1px solid #e6eaea;
  border-collapse: collapse;
  padding: 0px;
  cursor: pointer;
  line-height: 6vh;

  ${({ disabled }) => disabled && 'background-color: #e6eaea'};

  ${({ isWeekend }) => isWeekend && 'background-color: #eff0f0'};

  &:hover {
    background-color: #e6eaea;
  }
`;

export const EditableCell = ({
  value: initialValue = '',
  options,
  isWeekend = false,
  onClick,
  disabled,
}) => {
  const [value, setValue] = useState(initialValue);
  const [open, setOpen] = useState(false);

  useEffect(() => {
    setValue(initialValue);
  }, [initialValue]);

  const ref = useRef();

  const width = ref.current?.clientWidth || 0;

  useOnClickOutside(ref, () => setOpen(false));

  return (
    <EditableCellLayout
      onClick={() => setOpen((prev) => !prev)}
      ref={ref}
      isWeekend={isWeekend}
      disabled={disabled}
    >
      <CellText>{value}</CellText>
      {!disabled && (
        <Dropdown
          open={open}
          options={options}
          onClick={(e) => {
            onClick(e.target.getAttribute('value'));
            setValue(e.target.getAttribute('value'));
          }}
          width={width}
        />
      )}
    </EditableCellLayout>
  );
};
