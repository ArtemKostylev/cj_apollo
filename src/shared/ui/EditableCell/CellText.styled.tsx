import styled from 'styled-components';

export const CellText = styled.p`
  border: none;
  outline: none;
  width: 100%;
  height: 100%;
  margin: 0px;
`;