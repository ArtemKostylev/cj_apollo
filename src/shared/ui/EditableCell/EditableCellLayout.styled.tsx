import styled from 'styled-components';

export const EditableCellLayout = styled.td<{ disabled: boolean, isWeekend: boolean }>`
  border: 1px solid #e6eaea;
  border-top: none;
  border-collapse: collapse;
  padding: 0;
  cursor: pointer;
  line-height: 6vh;

  ${({disabled}) => disabled && 'background-color: #e6eaea'};

  ${({isWeekend}) => isWeekend && 'background-color: #eff0f0'};

  &:hover {
    background-color: #e6eaea;
  }
`;
