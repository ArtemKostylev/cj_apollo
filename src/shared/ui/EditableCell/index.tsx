import React, {useState, useEffect, useRef} from 'react';
import {useOnClickOutside} from '../../../hooks/useOnClickOutside';
import {Dropdown, DropdownOptionType} from '../Dropdown';
import {EditableCellLayout} from './EditableCellLayout.styled';
import {CellText} from './CellText.styled';

type Props = {
  value: string;
  options: Map<string, DropdownOptionType>;
  isWeekend: boolean;
  onClick: (value: string) => void;
  disabled: boolean;
}

export const EditableCell = ({value = '', options, isWeekend = false, onClick, disabled}: Props) => {
  const [dropdownValue, setDropdownValue] = useState(value);
  const [opened, setOpened] = useState(false);

  useEffect(() => {
    setDropdownValue(value);
  }, [value]);

  const ref = useRef() as any;

  const width = ref.current?.clientWidth || 0;

  useOnClickOutside(ref, () => setOpened(false));

  return (
    <EditableCellLayout onClick={() => setOpened((prev) => !prev)} ref={ref} isWeekend={isWeekend}
                        disabled={disabled}>
      <CellText>{dropdownValue === '.' ? 'âœ“' : dropdownValue}</CellText>
      {!disabled && (
        <Dropdown
          opened={opened}
          options={options}
          onSelect={(value) => {
            onClick(value);
            setDropdownValue(value);
          }}
          width={width}
        />
      )}
    </EditableCellLayout>
  );
};
