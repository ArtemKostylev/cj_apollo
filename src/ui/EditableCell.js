import React from 'react';
import { useOnClickOutside } from '../../scripts/utils';
import '../styles/Journal.css';

import styled from 'styled-components';

const DropdownOption = styled.li`
  display: flex;
  flex-direction: row;
  justify-content: space-evenly;
  text-align: center;
  padding: 0;
`;

const DropdownOptionsList = (options) => {

  const  = options.

};

const Dropdown = (visible) => {
  const onClick = (e) => {
    let mark = '';
    if (e.target.innerHTML !== 'пусто') mark = e.target.innerHTML;
    setValue(mark);
    setDropdownVisibility(false);
    if (!updateMyData(index, id, mark, group)) setValue('');
  };
  const [inverted, setInverted] = React.useState(false);

  const ref = React.useRef();

  React.useEffect(() => {
    if (dropdownVisible) {
      const bounds = ref.current.getBoundingClientRect();
      if (
        bounds.bottom >
        (window.innerHeight || document.documentElement.clientHeight)
      ) {
        setInverted(true);
      }
    }
  }, [setInverted, dropdownVisible]);

  return (
    <div
      className={`dropdown ${visible ? 'visible' : ''} ${
        !inverted || 'inverted'
      }`}
      style={{ width: `${dropdownWidth}px` }}
      ref={ref}
    >
      {
        <ul>
          <li onClick={onClick}>
            <p style={{ width: '100%' }}>пусто</p>
          </li>
          <li>
            <p onClick={onClick}>.</p>
            <p onClick={onClick}>2</p>
          </li>
          <li>
            <p onClick={onClick}>3</p>
            <p onClick={onClick}>4-</p>
          </li>
          <li>
            <p onClick={onClick}>4</p>
            <p onClick={onClick}>4+</p>
          </li>
          <li>
            <p onClick={onClick}>5-</p>
            <p onClick={onClick}>5</p>
          </li>
          <li>
            <p onClick={onClick}>5+</p>
            <p onClick={onClick}>Б</p>
          </li>
        </ul>
      }
    </div>
  );
};

// TODO: fix all import to named from default

export const EditableCell = ({
  value: initialValue,
  row: index,
  column: id,
  weekend = '',
  updateMyData,
  group = -1,
  disabled = false,
}) => {
  const [value, setValue] = React.useState(initialValue);
  const [dropdownVisible, setDropdownVisibility] = React.useState(false);
  const [dropdownWidth, setDropdownWidth] = React.useState(0);

  const ref = React.useRef();

  useOnClickOutside(ref, () => setDropdownVisibility(false));

  const onClick = () => {
    if (!dropdownVisible) setDropdownVisibility(true);
  };

  React.useEffect(() => {
    setValue(initialValue);
  }, [initialValue]);

  React.useEffect(() => {
    setDropdownWidth(ref?.current?.clientWidth || 0);
  }, [setDropdownWidth, ref?.current?.clientWidth]);

  React.useEffect(() => {});
  if (disabled) {
    return (
      <td ref={ref}>
        <p style={{ color: 'gray', margin: '0' }}>{value}</p>
      </td>
    );
  }
  return (
    <td onClick={onClick} ref={ref} className={`${weekend}`}>
      <p className='input'>{value}</p>
      {Dropdown(dropdownVisible)}
    </td>
  );
};
