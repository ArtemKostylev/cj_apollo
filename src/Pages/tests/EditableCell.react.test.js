import react from "react";
import renderer, { act } from "react-test-renderer";
import EditableCell from "../../components/EditableCell";

test("Drawer opens after click", () => {
  const component = renderer.create(
    <EditableCell value="" row={1} column={1} updateMyData={() => {}} />
  );
  let tree = component.toJSON();

  expect(tree).toMatchSnapshot();

  act(() => tree.props.onClick());
});
