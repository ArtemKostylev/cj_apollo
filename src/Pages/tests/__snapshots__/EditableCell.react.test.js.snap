// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drawer opens after click 1`] = `
<td
  className=""
  onClick={[Function]}
>
  <p
    className="input"
  >
    
  </p>
  <div
    className="dropdown  true"
    style={
      Object {
        "width": "0px",
      }
    }
  >
    <ul>
       
      <li
        onClick={[Function]}
      >
        <p
          style={
            Object {
              "width": "100%",
            }
          }
        >
          пусто
        </p>
      </li>
      <li>
        <p
          onClick={[Function]}
        >
          .
        </p>
        <p
          onClick={[Function]}
        >
          2
        </p>
      </li>
      <li>
        <p
          onClick={[Function]}
        >
          3
        </p>
        <p
          onClick={[Function]}
        >
          4-
        </p>
      </li>
      <li>
        <p
          onClick={[Function]}
        >
          4
        </p>
        <p
          onClick={[Function]}
        >
          4+
        </p>
      </li>
      <li>
        <p
          onClick={[Function]}
        >
          5-
        </p>
        <p
          onClick={[Function]}
        >
          5
        </p>
      </li>
      <li>
        <p
          onClick={[Function]}
        >
          5+
        </p>
        <p
          onClick={[Function]}
        >
          Б
        </p>
      </li>
    </ul>
  </div>
</td>
`;
